### PR TITLE
ignition: add support for ignition config file

### LINKED ISSUE #124: Add `--ignition config.ign` support
This would simplify use of generic VM images which can be configured using ignition.

The podman code doing this is https://github.com/containers/podman/blob/6487940534c1065d6c7753e3b6dfbe253666537d/pkg/machine/applehv/ignition.go

### PR DESCRIPTION
**Pull request by lstocchi**
it adds support to pass an ignition config file to vfkit that handles the provisioning process.

it resolves #124

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -191,6 +202,39 @@ func (vm *VirtualMachine) TimeSync() *TimeSync {
...
+	if cmdlineOpts == "" {
+		return nil
+	}
+	opts := strings.Split(cmdlineOpts, ",")
+	if len(opts) != 1 {
+		return fmt.Errorf("ignition only accept one option in command line argument")
+	}
+
+	socketPath := filepath.Join(os.TempDir(), "ignition.sock")
+	dev, err := VirtioVsockNew(1024, socketPath, true)
```
Comment (baude) [inline pkg/config/config.go:225]: pop to const ?

Comment (baude) [reply pkg/config/config.go:225]: and maybe line above?

File: cmd/vfkit/main.go [hunk 2]
```diff
@@ -86,6 +91,10 @@ func newVMConfiguration(opts *cmdline.Options) (*config.VirtualMachine, error) {
 		return nil, err
 	}
 
+	if err := vmConfig.AddIgnitionFileFromCmdLine(opts.IgnitionPath); err != nil {
+		return nil, err
```
Comment (baude) [inline cmd/vfkit/main.go:95]: perhaps consider wrapping this here to make it meaningful

File: cmd/vfkit/main.go [hunk 3]
```diff
@@ -198,3 +216,37 @@ func runVirtualMachine(vmConfig *config.VirtualMachine, vm *vf.VirtualMachine) e
 
 	return <-errCh
 }
+
+func startServer(ignitionFile string, ignitionSocketPath string) error {
```
Comment (baude) [inline cmd/vfkit/main.go:220]: consider passing ignitionFile as a reader

**Comment by baude** [comment]
flyby review, nice work work @lstocchi ... all my comments are largely things to think about, neat idea too.

**Comment by lstocchi** [comment]
@baude thanks for the suggestions. Hopefully I got all of them. 👍 

Gonna add some tests tomorrow morning so that it can be reviewed again

**Comment by baude** [comment]
@lstocchi being that ignition is on the rarer side of things, I wonder you should create `contrib/ignition` and drop something of an example in there ?  I assume it is in JSON so not something you could heavily comment on but a README.md in that dir could explain it ?  This is extra credit of course :)

**Review by baude — APPROVED** [review]
Just the proposal to add an example file comment ... lgtm otherwise

File: pkg/config/json_test.go [hunk 4]
```diff
@@ -85,6 +85,16 @@ var jsonTests = map[string]jsonTest{
 		},
 		expectedJSON: `{"vcpus":3,"memoryBytes":4194304000,"bootloader":{"kind":"linuxBootloader","vmlinuzPath":"/vmlinuz","initrdPath":"/initrd","kernelCmdLine":"console=hvc0"},"timesync":{"vsockPort":1234}}`,
 	},
+	"TestIgnition": {
```
Comment (cfergeau) [inline pkg/config/json_test.go:88]: These `json_test` changes could be squashed in the previous commit.

Comment (lstocchi) [reply pkg/config/json_test.go:88]: updated 👍

File: cmd/vfkit/main.go [hunk 5]
```diff
@@ -198,3 +220,33 @@ func runVirtualMachine(vmConfig *config.VirtualMachine, vm *vf.VirtualMachine) e
 
 	return <-errCh
 }
+
+func startIgnitionProvisionerServer(ignitionReader io.Reader, ignitionSocketPath string) error {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/", func(w http.ResponseWriter, _ *http.Request) {
+		_, err := io.Copy(w, ignitionReader)
```
Comment (cfergeau) [inline cmd/vfkit/main.go:227]: This could use https://pkg.go.dev/net/http#ServeContent (no need to make this change right, this can be a follow-up)

**Review by cfergeau — APPROVED** [review]
/lgtm
/approve

**Pull request merged**